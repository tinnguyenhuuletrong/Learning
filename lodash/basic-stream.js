function Observable(cb) {
  const subscrible = next => cb({ next })
  return { subscrible }
}

const source$ = new Observable(obs => {
  let count = 1
  setInterval(() => obs.next(count++), 1000)
})

source$.subscrible(x => console.log('sub1', x))
source$.subscrible(x => console.log('sub2', x))
