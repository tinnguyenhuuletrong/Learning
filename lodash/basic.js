const lo = require('lodash')

function arrayMethods() {
  const a = [1, 2, 3, 4, 5]
  const b = [1, 3]
  console.log('difference:', a, b, '->', lo.difference(a, b))
  console.log('drop', a, '->', lo.drop(a, 2))
  console.log('fromPairs', lo.fromPairs([['a', 1], ['b', 2]]))
  console.log('intersection', '->', lo.intersection([2, 1], [2, 3]))
  console.log(
    'binary search sorted index',
    [30, 50, 90, 100],
    40,
    '->',
    lo.sortedIndex([30, 50, 90, 100], 40)
  )

  const p3 = [['a', 'b', 'c'], [1, 2, 3], [{ e: 1 }, { e: 2 }, { e: 3 }]]
  console.log('zip', p3[0], p3[1], p3[2], '\n->', lo.zip(...p3))

  console.log(
    'zipWith sum',
    [1, 2],
    [10, 20],
    [100, 200],
    '->',
    lo.zipWith([1, 2], [10, 20], [100, 200], (a, b, c) => a + b + c)
  )
}

function collectionMethods() {
  console.log('countBy', lo.countBy(['a', 'b', 'c', 'd', 'a']))
  console.log('flatMap', lo.flatMap([1, 2], a => [a, a]))

  // GroupBy
  const p = [
    {
      name: 'a',
      cat: '1'
    },
    {
      name: 'b',
      cat: '1'
    },
    {
      name: 'c',
      cat: '2'
    }
  ]
  console.log('groupBy', lo.groupBy(p, 'cat'))

  // Sort by `user` in ascending order and by `age` in descending order.
  const users = [
    { user: 'fred', age: 48 },
    { user: 'barney', age: 34 },
    { user: 'fred', age: 40 },
    { user: 'barney', age: 36 }
  ]
  console.log('orderBy', lo.orderBy(users, ['user', 'age'], ['asc', 'desc']))
}

function objectMethods() {
  var object = {
    a: [
      {
        b: { c: 3 }
      },
      4
    ],
    x: 'hello'
  }

  console.log(
    'object at (return array):',
    '[a[0].b.c, a[1]]',
    '->',
    lo.at(object, ['a[0].b.c', 'a[1]'])
  )
  console.log('object at (return array):', 'x', '->', lo.at(object, 'x'))
}

objectMethods()
