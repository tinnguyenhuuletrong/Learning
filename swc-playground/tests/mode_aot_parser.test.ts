import * as swc from "@swc/core";
import { isArray, isFunction } from "lodash";
import {
  TLiteAotCompileVisitor,
  CompilerContext,
} from "../src/TLiteAotCompileVisitor";

describe("TLite aot compiler: expression, assignment", () => {
  test("1 + (3 * 5)**2", async () => {
    const exp = `1 + (3 * 5)**2;`;
    const astTree = await swc.parse(exp);
    const runtime = new TLiteAotCompileVisitor();

    const ctx = new CompilerContext();
    const finalVal = runtime.run(astTree, ctx);
    expect(runtime).toMatchSnapshot();
    console.log(exp);
    console.log(ctx.toString());
  });

  test("res = 'hello'", async () => {
    const exp = `
      res = 'hello'
      res = 'hi'
    `;
    const astTree = await swc.parse(exp);
    const runtime = new TLiteAotCompileVisitor();

    const ctx = new CompilerContext();
    runtime.run(astTree, ctx);
    expect(runtime).toMatchSnapshot();
    console.log(exp);
    console.log(ctx.toString());
  });

  test("res = 1 + (3 * 5)/2", async () => {
    const exp = `res = 1 + (3 * 5)/2;`;
    const astTree = await swc.parse(exp);
    const runtime = new TLiteAotCompileVisitor();

    const ctx = new CompilerContext();
    runtime.run(astTree, ctx);
    expect(runtime).toMatchSnapshot();
    console.log(exp);
    console.log(ctx.toString());
  });

  test("res = inp > 2", async () => {
    const exp = `res = inp > 2`;
    const astTree = await swc.parse(exp);
    const runtime = new TLiteAotCompileVisitor();
    const ctx = new CompilerContext();
    runtime.run(astTree, ctx);
    expect(runtime).toMatchSnapshot();
    console.log(exp);
    console.log(ctx.toString());
  });

  test("res = inp <= 2", async () => {
    const exp = `res = inp <= 2`;
    const astTree = await swc.parse(exp);
    const runtime = new TLiteAotCompileVisitor();
    const ctx = new CompilerContext();
    runtime.run(astTree, ctx);
    expect(runtime).toMatchSnapshot();
    console.log(exp);
    console.log(ctx.toString());
  });

  test("res = inp > 2 && inp <= 5", async () => {
    const exp = `res = inp > 2 && inp <= 5`;
    const astTree = await swc.parse(exp);
    const runtime = new TLiteAotCompileVisitor();
    const ctx = new CompilerContext();
    runtime.run(astTree, ctx);
    expect(runtime).toMatchSnapshot();
    console.log(exp);
    console.log(ctx.toString());
  });

  test("res = inp < 2 || inp > 5", async () => {
    const exp = `res = inp < 2 || inp > 5`;
    const astTree = await swc.parse(exp);
    const runtime = new TLiteAotCompileVisitor();
    const ctx = new CompilerContext();
    runtime.run(astTree, ctx);
    expect(runtime).toMatchSnapshot();
    console.log(exp);
    console.log(ctx.toString());
  });

  test("res = !(+inp < 2 || +inp > 5)", async () => {
    const exp = `res = !(+inp < 2 || +inp > 5)`;
    const astTree = await swc.parse(exp);
    const runtime = new TLiteAotCompileVisitor();
    const ctx = new CompilerContext();
    runtime.run(astTree, ctx);
    expect(runtime).toMatchSnapshot();
    console.log(exp);
    console.log(ctx.toString());
  });
});

describe("TLite aot compiler: array, object", () => {
  test("res = [1,2,inp > 18,'name']", async () => {
    const exp = `res = [1,2,inp > 18,'name']`;
    const astTree = await swc.parse(exp);
    const runtime = new TLiteAotCompileVisitor();
    const ctx = new CompilerContext();
    runtime.run(astTree, ctx);
    expect(runtime).toMatchSnapshot();
    console.log(exp);
    console.log(ctx.toString());
  });

  test("res = [1,2,{c:true}]", async () => {
    const exp = `res = [1,2,{c:true}]`;
    const astTree = await swc.parse(exp);
    const runtime = new TLiteAotCompileVisitor();
    const ctx = new CompilerContext();
    runtime.run(astTree, ctx);
    expect(runtime).toMatchSnapshot();
    console.log(exp);
    console.log(ctx.toString());
  });

  test("res = {b:true, i: 1, v: inp, a: [1,3], o: {nestest: {}}}", async () => {
    const exp = `res = {b:true, i: 1, v: inp, a: [1,3], o: {nestest: {}}}`;
    const astTree = await swc.parse(exp);
    const runtime = new TLiteAotCompileVisitor();
    const ctx = new CompilerContext();
    runtime.run(astTree, ctx);
    expect(runtime).toMatchSnapshot();
    console.log(exp);
    console.log(ctx.toString());
  });
});
