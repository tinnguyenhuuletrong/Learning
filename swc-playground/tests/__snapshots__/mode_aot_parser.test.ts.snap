// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TLite aot compiler: array, object res = [1,2,{c:true}] 1`] = `
TLiteAotCompileVisitor {
  "ctx": CompilerContext {
    "ops": Array [
      Op {
        "op": "MSAVE",
        "params": Object {
          "nextVal": Op {
            "op": "SVAL",
            "params": Object {
              "type": "array",
              "v": Array [
                Op {
                  "op": "SVAL",
                  "params": Object {
                    "type": "number",
                    "v": 1,
                  },
                },
                Op {
                  "op": "SVAL",
                  "params": Object {
                    "type": "number",
                    "v": 2,
                  },
                },
                Op {
                  "op": "SVAL",
                  "params": Object {
                    "type": "object",
                    "v": Array [
                      Object {
                        "key": "c",
                        "val": Op {
                          "op": "SVAL",
                          "params": Object {
                            "type": "bool",
                            "v": true,
                          },
                        },
                      },
                    ],
                  },
                },
              ],
            },
          },
          "path": "res",
          "perform": "=",
        },
      },
    ],
  },
}
`;

exports[`TLite aot compiler: array, object res = [1,2,inp > 18,'name'] 1`] = `
TLiteAotCompileVisitor {
  "ctx": CompilerContext {
    "ops": Array [
      Op {
        "op": "MSAVE",
        "params": Object {
          "nextVal": Op {
            "op": "SVAL",
            "params": Object {
              "type": "array",
              "v": Array [
                Op {
                  "op": "SVAL",
                  "params": Object {
                    "type": "number",
                    "v": 1,
                  },
                },
                Op {
                  "op": "SVAL",
                  "params": Object {
                    "type": "number",
                    "v": 2,
                  },
                },
                Op {
                  "op": "BEXP",
                  "params": Object {
                    "op": ">",
                    "v1": Op {
                      "op": "SVAL",
                      "params": Object {
                        "type": "variable",
                        "v": "inp",
                      },
                    },
                    "v2": Op {
                      "op": "SVAL",
                      "params": Object {
                        "type": "number",
                        "v": 18,
                      },
                    },
                  },
                },
                Op {
                  "op": "SVAL",
                  "params": Object {
                    "type": "string",
                    "v": "name",
                  },
                },
              ],
            },
          },
          "path": "res",
          "perform": "=",
        },
      },
    ],
  },
}
`;

exports[`TLite aot compiler: array, object res = {b:true, i: 1, v: inp, a: [1,3], o: {nestest: {}}} 1`] = `
TLiteAotCompileVisitor {
  "ctx": CompilerContext {
    "ops": Array [
      Op {
        "op": "MSAVE",
        "params": Object {
          "nextVal": Op {
            "op": "SVAL",
            "params": Object {
              "type": "object",
              "v": Array [
                Object {
                  "key": "b",
                  "val": Op {
                    "op": "SVAL",
                    "params": Object {
                      "type": "bool",
                      "v": true,
                    },
                  },
                },
                Object {
                  "key": "i",
                  "val": Op {
                    "op": "SVAL",
                    "params": Object {
                      "type": "number",
                      "v": 1,
                    },
                  },
                },
                Object {
                  "key": "v",
                  "val": Op {
                    "op": "SVAL",
                    "params": Object {
                      "type": "variable",
                      "v": "inp",
                    },
                  },
                },
                Object {
                  "key": "a",
                  "val": Op {
                    "op": "SVAL",
                    "params": Object {
                      "type": "array",
                      "v": Array [
                        Op {
                          "op": "SVAL",
                          "params": Object {
                            "type": "number",
                            "v": 1,
                          },
                        },
                        Op {
                          "op": "SVAL",
                          "params": Object {
                            "type": "number",
                            "v": 3,
                          },
                        },
                      ],
                    },
                  },
                },
                Object {
                  "key": "o",
                  "val": Op {
                    "op": "SVAL",
                    "params": Object {
                      "type": "object",
                      "v": Array [
                        Object {
                          "key": "nestest",
                          "val": Op {
                            "op": "SVAL",
                            "params": Object {
                              "type": "object",
                              "v": Array [],
                            },
                          },
                        },
                      ],
                    },
                  },
                },
              ],
            },
          },
          "path": "res",
          "perform": "=",
        },
      },
    ],
  },
}
`;

exports[`TLite aot compiler: expression, assignment 1 + (3 * 5)**2 1`] = `
TLiteAotCompileVisitor {
  "ctx": CompilerContext {
    "ops": Array [
      Op {
        "op": "BEXP",
        "params": Object {
          "op": "+",
          "v1": Op {
            "op": "SVAL",
            "params": Object {
              "type": "number",
              "v": 1,
            },
          },
          "v2": Op {
            "op": "BEXP",
            "params": Object {
              "op": "**",
              "v1": Op {
                "op": "BEXP",
                "params": Object {
                  "op": "*",
                  "v1": Op {
                    "op": "SVAL",
                    "params": Object {
                      "type": "number",
                      "v": 3,
                    },
                  },
                  "v2": Op {
                    "op": "SVAL",
                    "params": Object {
                      "type": "number",
                      "v": 5,
                    },
                  },
                },
              },
              "v2": Op {
                "op": "SVAL",
                "params": Object {
                  "type": "number",
                  "v": 2,
                },
              },
            },
          },
        },
      },
    ],
  },
}
`;

exports[`TLite aot compiler: expression, assignment res = !(+inp < 2 || +inp > 5) 1`] = `
TLiteAotCompileVisitor {
  "ctx": CompilerContext {
    "ops": Array [
      Op {
        "op": "MSAVE",
        "params": Object {
          "nextVal": Op {
            "op": "UEXP",
            "params": Object {
              "args": Array [
                Op {
                  "op": "BEXP",
                  "params": Object {
                    "op": "||",
                    "v1": Op {
                      "op": "BEXP",
                      "params": Object {
                        "op": "<",
                        "v1": Op {
                          "op": "UEXP",
                          "params": Object {
                            "args": Array [
                              Op {
                                "op": "SVAL",
                                "params": Object {
                                  "type": "variable",
                                  "v": "inp",
                                },
                              },
                            ],
                            "op": "+",
                          },
                        },
                        "v2": Op {
                          "op": "SVAL",
                          "params": Object {
                            "type": "number",
                            "v": 2,
                          },
                        },
                      },
                    },
                    "v2": Op {
                      "op": "BEXP",
                      "params": Object {
                        "op": ">",
                        "v1": Op {
                          "op": "UEXP",
                          "params": Object {
                            "args": Array [
                              Op {
                                "op": "SVAL",
                                "params": Object {
                                  "type": "variable",
                                  "v": "inp",
                                },
                              },
                            ],
                            "op": "+",
                          },
                        },
                        "v2": Op {
                          "op": "SVAL",
                          "params": Object {
                            "type": "number",
                            "v": 5,
                          },
                        },
                      },
                    },
                  },
                },
              ],
              "op": "!",
            },
          },
          "path": "res",
          "perform": "=",
        },
      },
    ],
  },
}
`;

exports[`TLite aot compiler: expression, assignment res = 'hello' 1`] = `
TLiteAotCompileVisitor {
  "ctx": CompilerContext {
    "ops": Array [
      Op {
        "op": "MSAVE",
        "params": Object {
          "nextVal": Op {
            "op": "SVAL",
            "params": Object {
              "type": "string",
              "v": "hello",
            },
          },
          "path": "res",
          "perform": "=",
        },
      },
      Op {
        "op": "MSAVE",
        "params": Object {
          "nextVal": Op {
            "op": "SVAL",
            "params": Object {
              "type": "string",
              "v": "hi",
            },
          },
          "path": "res",
          "perform": "=",
        },
      },
    ],
  },
}
`;

exports[`TLite aot compiler: expression, assignment res = 1 + (3 * 5)/2 1`] = `
TLiteAotCompileVisitor {
  "ctx": CompilerContext {
    "ops": Array [
      Op {
        "op": "MSAVE",
        "params": Object {
          "nextVal": Op {
            "op": "BEXP",
            "params": Object {
              "op": "+",
              "v1": Op {
                "op": "SVAL",
                "params": Object {
                  "type": "number",
                  "v": 1,
                },
              },
              "v2": Op {
                "op": "BEXP",
                "params": Object {
                  "op": "/",
                  "v1": Op {
                    "op": "BEXP",
                    "params": Object {
                      "op": "*",
                      "v1": Op {
                        "op": "SVAL",
                        "params": Object {
                          "type": "number",
                          "v": 3,
                        },
                      },
                      "v2": Op {
                        "op": "SVAL",
                        "params": Object {
                          "type": "number",
                          "v": 5,
                        },
                      },
                    },
                  },
                  "v2": Op {
                    "op": "SVAL",
                    "params": Object {
                      "type": "number",
                      "v": 2,
                    },
                  },
                },
              },
            },
          },
          "path": "res",
          "perform": "=",
        },
      },
    ],
  },
}
`;

exports[`TLite aot compiler: expression, assignment res = inp < 2 || inp > 5 1`] = `
TLiteAotCompileVisitor {
  "ctx": CompilerContext {
    "ops": Array [
      Op {
        "op": "MSAVE",
        "params": Object {
          "nextVal": Op {
            "op": "BEXP",
            "params": Object {
              "op": "||",
              "v1": Op {
                "op": "BEXP",
                "params": Object {
                  "op": "<",
                  "v1": Op {
                    "op": "SVAL",
                    "params": Object {
                      "type": "variable",
                      "v": "inp",
                    },
                  },
                  "v2": Op {
                    "op": "SVAL",
                    "params": Object {
                      "type": "number",
                      "v": 2,
                    },
                  },
                },
              },
              "v2": Op {
                "op": "BEXP",
                "params": Object {
                  "op": ">",
                  "v1": Op {
                    "op": "SVAL",
                    "params": Object {
                      "type": "variable",
                      "v": "inp",
                    },
                  },
                  "v2": Op {
                    "op": "SVAL",
                    "params": Object {
                      "type": "number",
                      "v": 5,
                    },
                  },
                },
              },
            },
          },
          "path": "res",
          "perform": "=",
        },
      },
    ],
  },
}
`;

exports[`TLite aot compiler: expression, assignment res = inp <= 2 1`] = `
TLiteAotCompileVisitor {
  "ctx": CompilerContext {
    "ops": Array [
      Op {
        "op": "MSAVE",
        "params": Object {
          "nextVal": Op {
            "op": "BEXP",
            "params": Object {
              "op": "<=",
              "v1": Op {
                "op": "SVAL",
                "params": Object {
                  "type": "variable",
                  "v": "inp",
                },
              },
              "v2": Op {
                "op": "SVAL",
                "params": Object {
                  "type": "number",
                  "v": 2,
                },
              },
            },
          },
          "path": "res",
          "perform": "=",
        },
      },
    ],
  },
}
`;

exports[`TLite aot compiler: expression, assignment res = inp > 2 && inp <= 5 1`] = `
TLiteAotCompileVisitor {
  "ctx": CompilerContext {
    "ops": Array [
      Op {
        "op": "MSAVE",
        "params": Object {
          "nextVal": Op {
            "op": "BEXP",
            "params": Object {
              "op": "&&",
              "v1": Op {
                "op": "BEXP",
                "params": Object {
                  "op": ">",
                  "v1": Op {
                    "op": "SVAL",
                    "params": Object {
                      "type": "variable",
                      "v": "inp",
                    },
                  },
                  "v2": Op {
                    "op": "SVAL",
                    "params": Object {
                      "type": "number",
                      "v": 2,
                    },
                  },
                },
              },
              "v2": Op {
                "op": "BEXP",
                "params": Object {
                  "op": "<=",
                  "v1": Op {
                    "op": "SVAL",
                    "params": Object {
                      "type": "variable",
                      "v": "inp",
                    },
                  },
                  "v2": Op {
                    "op": "SVAL",
                    "params": Object {
                      "type": "number",
                      "v": 5,
                    },
                  },
                },
              },
            },
          },
          "path": "res",
          "perform": "=",
        },
      },
    ],
  },
}
`;

exports[`TLite aot compiler: expression, assignment res = inp > 2 1`] = `
TLiteAotCompileVisitor {
  "ctx": CompilerContext {
    "ops": Array [
      Op {
        "op": "MSAVE",
        "params": Object {
          "nextVal": Op {
            "op": "BEXP",
            "params": Object {
              "op": ">",
              "v1": Op {
                "op": "SVAL",
                "params": Object {
                  "type": "variable",
                  "v": "inp",
                },
              },
              "v2": Op {
                "op": "SVAL",
                "params": Object {
                  "type": "number",
                  "v": 2,
                },
              },
            },
          },
          "path": "res",
          "perform": "=",
        },
      },
    ],
  },
}
`;
